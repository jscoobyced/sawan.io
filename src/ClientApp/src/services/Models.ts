import { Url } from "url";

export enum Language {
    English,
    French
}

export interface Link {
    text: string;
    url: string;
    target?: string;
    title?: string;
}

export interface AllContent {
    navigationMenuContent: NavigationMenuContent;
    homeContent: HomeContent;
    menuContent: MenuContent;
    footerContent: FooterContent;
}

export interface NavigationMenuContent {
    home: string | string[];
    websiteName: string | string[];
    applications: string | string[];
    cryptoCurrency: string | string[];
    healthMonitor: string | string[];
    about: string | string[];
}

export interface HomeContent {
    articleTitle: string | string[];
    article: string | string[];
}

export interface MenuContent {
    links: Link[];
}

export interface FooterContent {
    copyright: string | string[];
    credits: string | string[];
    year: string | string[];
}