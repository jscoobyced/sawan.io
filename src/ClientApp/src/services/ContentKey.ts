export class ContentKey {

    // General
    public static WebsiteName = 'WEBSITENAME';

    // Navigation Menu
    public static Home = 'HOME';
    public static About = 'ABOUT';
    public static Applications = 'APPLICATIONS';
    public static CryptoCurrency = 'CRYPTOCURRENCY';
    public static HealthMonitor = 'HEALTHMONITOR';

    // Home
    public static MainArticleTitle = 'MAINARTICLETITLE';
    public static MainArticle = 'MAINARTICLE';

    // Footer
    public static FooterCopyright = 'FOOTERCOPYRIGHT';
    public static FooterCredits = 'FOOTERCREDITS';
    public static FooterYear = 'FOOTERYEAR';
}
