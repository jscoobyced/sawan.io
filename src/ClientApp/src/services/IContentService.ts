import { BlogPage, Language, MainContent } from './Models';

export interface IContentService {
    getMainContent(language: Language): Promise<MainContent>;
    getDefaultMainContent(): MainContent;
    getBlogPage(maxResult: number, from: number): Promise<BlogPage>;
    getDefaultBlogPage(): BlogPage;
}
