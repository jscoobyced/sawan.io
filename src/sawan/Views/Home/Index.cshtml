@{
    ViewData["Title"] = "Home Page";
}

@section styles {
    <link rel="stylesheet" href="~/style.9eeba86c304bf4dc2e7d.css?a419536a7db2745d4e3b">
    <link rel="icon" href="/favicon.ico" type="image/x-icon">

}
<div id="app-root">Something awesome is coming...</div>

@section scripts {
    <script src="~/vendor.71b3b645b62e3a4154f0.js?a419536a7db2745d4e3b"></script>
    <script src="~/main.e8a8402b2811a0f34ea4.js?a419536a7db2745d4e3b"></script>
}
