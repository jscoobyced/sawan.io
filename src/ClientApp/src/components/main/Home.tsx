import React from 'react';
import { NavLink } from 'react-router-dom';
import xss from 'xss';
import { HomeContent } from '../../services/Models';

export interface HomeProps {
    homeContent: HomeContent;
}

export class Home extends React.Component<HomeProps> {

    private readonly options = {
        whiteList: {
            a: ['href', 'title', 'target'],
            br: []
        }
    };
    private readonly xssFilter = new xss.FilterXSS(this.options);

    public constructor(props: HomeProps) {
        super(props);
    }

    public render() {
        const articles: string[] = this.props.homeContent.article as string[];
        const titles: string[] = this.props.homeContent.articleTitle as string[];
        if (!articles || !titles || articles.length !== titles.length) {
            return <span></span>;
        }

        return articles.map((article, key) => <article key={key}>
            <h1>{titles[key]}</h1>
            <div>
                <span dangerouslySetInnerHTML={{ __html: this.getSafeContent(article) }}></span>
                <NavLink to='/' className='link-article'>Read more...</NavLink>
            </div>
        </article>);
    }

    private getSafeContent(content: string): string {
        const safeContent = this.xssFilter.process(content);
        let safeSummary = safeContent.substring(0, 1000);
        if (safeContent !== safeSummary) {
            safeSummary = safeSummary.substring(0, safeSummary.lastIndexOf(" "));
            safeSummary += '&#8230;';
        }
        return safeSummary;
    }
}
