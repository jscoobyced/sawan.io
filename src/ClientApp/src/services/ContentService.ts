import { ContentKey } from './ContentKey';
import { AllContent, Language } from './Models';

export class ContentService {

    private readonly content: Array<{ [key: string]: string | string[]; }> = [];

    public updateContent(language: Language): AllContent {
        this.content[language] = {};
        this.content[language][ContentKey.WebsiteName] = 'sawan.io';
        this.content[language][ContentKey.Home] = 'Home';
        this.content[language][ContentKey.About] = 'About';
        this.content[language][ContentKey.Applications] = 'Applications';
        this.content[language][ContentKey.CryptoCurrency] = 'Cryptocurrency';
        this.content[language][ContentKey.HealthMonitor] = 'Health Monitor';
        const date = '2018-11-15';
        const displayDate = (new Date(date).toISOString().substring(0, 10));
        this.content[language][ContentKey.MainArticleTitle] = [`Latest news - ${displayDate}`,
            'Lorem Ipsum', 'Some other news'];
        this.content[language][ContentKey.MainArticle] = ['Grand opening of sawan.io.<br />This website is about my'
            + ' personal training on various web-technologies: TypeScript, webpack, ReactJS, Secure HTTP headers...'
            + ' It includes a simple (and a bit naive) crypto-currency analyser: it can suggest you to sell'
            + ' or buy. Use at your own risk, it is amateur predictions, you have been warned.'
            + '<br />There will be also an application to track health metrics (weight, % body fat, vascular'
            + ' fat ratio...). Or maybe not, this might be my next project.<br /> This project is open-sourced on <a'
            + ' target=\'_blank\' href=\'https://github.com/jscoobyced/sawan.io\'>Github</a>. Enjoy and have fun.',
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore'
        + ' et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut'
        + ' aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse'
        + ' cillum dolore eu fugiat nulla pariatur.<br />Excepteur sint occaecat cupidatat non proident, sunt in'
        + ' culpa qui officia deserunt mollit anim id est laborum. Commodo quis imperdiet massa tincidunt.'
        + ' Non enim praesent elementum facilisis leo. Adipiscing diam donec adipiscing tristique. Velit'
        + ' euismod in pellentesque massa placerat. Interdum velit euismod in pellentesque. Phasellus'
        + ' vestibulum lorem sed risus ultricies tristique nulla. Egestas erat imperdiet sed euismod nisi'
        + ' porta lorem mollis aliquam. Nulla facilisi etiam dignissim diam quis enim lobortis scelerisque'
        + ' fermentum. Morbi tristique senectus et netus et malesuada fames.<br />Nisi vitae suscipit tellus'
        + ' mauris a diam maecenas. Sed risus ultricies tristique nulla aliquet. Sagittis eu volutpat odio'
        + ' facilisis mauris sit amet. Tempor orci eu lobortis elementum nibh tellus molestie nunc non.'
        + ' Turpis nunc eget lorem dolor.',
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore'
        + ' et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut'
        + ' aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse'
        + ' cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in'
        + ' culpa qui officia deserunt mollit anim id est laborum. Commodo quis imperdiet massa tincidunt.'
        + ' Non enim praesent elementum facilisis leo. Adipiscing diam donec adipiscing tristique. Velit'
        + ' euismod in pellentesque massa placerat. Interdum velit euismod in pellentesque. Phasellus'
        + ' vestibulum lorem sed risus ultricies tristique nulla. Egestas erat imperdiet sed euismod nisi'
        + ' porta lorem mollis aliquam. Nulla facilisi etiam dignissim diam quis enim lobortis scelerisque'
        + ' fermentum. Morbi tristique senectus et netus et malesuada fames.<br />Nisi vitae suscipit tellus'
        + ' mauris a diam maecenas. Sed risus ultricies tristique nulla aliquet. Sagittis eu volutpat odio'
        + ' facilisis mauris sit amet. Tempor orci eu lobortis elementum nibh tellus molestie nunc non.'
        + ' Turpis nunc eget lorem dolor. Duis aute irure dolor in reprehenderit in voluptate velit esse'
        + ' cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in'
        + ' culpa qui officia deserunt mollit anim id est laborum. Commodo quis imperdiet massa tincidunt.'
        + ' Non enim praesent elementum facilisis leo. Adipiscing diam donec adipiscing tristique. Velit'
        + ' euismod in pellentesque massa placerat. Interdum velit euismod in pellentesque. Phasellus'
        + ' vestibulum lorem sed risus ultricies tristique nulla.<br />Egestas erat imperdiet sed euismod nisi'
        + ' porta lorem mollis aliquam. Nulla facilisi etiam dignissim diam quis enim lobortis scelerisque'
        + ' fermentum. Morbi tristique senectus et netus et malesuada fames. Nisi vitae suscipit tellus'
        + ' mauris a diam maecenas. Sed risus ultricies tristique nulla aliquet.'];
        this.content[language][ContentKey.FooterCopyright] = 'Copyright';
        this.content[language][ContentKey.FooterCredits] = 'sawan.io';
        this.content[language][ContentKey.FooterYear] = '2018';

        return this.updateAllContent(language);
    }

    public defaultContent(): AllContent {
        return {
            navigationMenuContent: {
                about: '',
                applications: '',
                cryptoCurrency: '',
                healthMonitor: '',
                home: '',
                websiteName: '',
            },
            homeContent: {
                article: [''],
                articleTitle: ['']
            },
            menuContent: {
                links: [{
                    text: '',
                    url: '',
                    target: '',
                    title: ''
                }]
            },
            footerContent: {
                credits: '',
                copyright: '',
                year: ''
            }
        };

    }

    private updateAllContent(language: Language): AllContent {
        return {
            navigationMenuContent: {
                about: this.content[language][ContentKey.About],
                applications: this.content[language][ContentKey.Applications],
                cryptoCurrency: this.content[language][ContentKey.CryptoCurrency],
                healthMonitor: this.content[language][ContentKey.HealthMonitor],
                home: this.content[language][ContentKey.Home],
                websiteName: this.content[language][ContentKey.WebsiteName],
            },
            homeContent: {
                article: this.content[language][ContentKey.MainArticle],
                articleTitle: this.content[language][ContentKey.MainArticleTitle]
            },
            menuContent: {
                links: [{
                    text: this.content[language][ContentKey.MainArticleTitle][0],
                    url: '#'
                }, {
                    text: this.content[language][ContentKey.MainArticleTitle][1],
                    url: '#'
                }, {
                    text: this.content[language][ContentKey.MainArticleTitle][2],
                    url: '#'
                }]
            },
            footerContent: {
                credits: this.content[language][ContentKey.FooterCredits],
                copyright: this.content[language][ContentKey.FooterCopyright],
                year: this.content[language][ContentKey.FooterYear]
            }
        };
    }
}
