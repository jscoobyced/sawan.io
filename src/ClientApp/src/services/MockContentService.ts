import { ContentService } from './ContentService';
import { BlogElement, BlogPage, Language, MainContent } from './Models';

export class MockContentService extends ContentService {

    private readonly data = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do'
        + ' ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco'
        + ' aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit '
        + ' cillum dolore eu fugiat nulla pariatur.<br />Excepteur sint occaecat cupidatat non proident, '
        + ' culpa qui officia deserunt mollit anim id est laborum. Commodo qui imperdiet massa tincidunt.'
        + ' Non enim praesent elementum facilisis leo. Adipiscing diam donec adipiscing tristique. Velit'
        + ' euismod in pellentesque massa placerat. Interdum velit euismod in pellentesque. Phasellus'
        + ' vestibulum lorem sed risus ultricies tristique nulla. Egestas erat imperdiet sed euismod nisi'
        + ' porta lorem mollis aliquam. Nulla facilisi etiam dignissim diam quis enim lobortis scelerisque'
        + ' fermentum. Morbi tristique senectus et netus et malesuada fames.<br />Vitae suscipit tellus'
        + ' mauris a diam maecenas. Sed risus ultricies tristique nulla aliquet. Sagittis eu volutpat odio'
        + ' facilisis mauris sit amet. Tempor orci eu lobortis elementum nibh tellus molestie nunc non.'
        + ' Turpis nunc eget lorem dolor.';

    protected updateMainContent(language: Language): Promise<MainContent> {
        const content: MainContent = {
            menuContent: {
                links: []
            },
            navigationMenuContent: {
                about: 'About',
                applications: 'Applications',
                cryptoCurrency: 'Crypto-Currency',
                healthMonitor: 'Health Monitor',
                home: 'Home',
                websiteName: 'sawan.io'
            },
            footerContent: {
                copyright: 'Copyright',
                credits: 'sawan.io',
                year: '2018 - ' + new Date().getFullYear()
            }
        };
        return Promise.resolve(content);
    }

    protected updateBlogPage(maxResult: number): Promise<BlogPage> {
        const date = '2018-11-15';
        const displayDate = (new Date(date).toISOString().substring(0, 10));
        const articles: BlogElement[] = [
            {
                article: `Latest news - ${displayDate}`,
                articleTitle: 'Grand opening of sawan.io.<br />This website is about my'
                    + ' personal training on various web-technologies: TypeScript, webpack, ReactJS, Secure HTTP'
                    + '  headers... It includes a simple (and a bit naive) crypto-currency analyser: it can'
                    + ' suggest you to sell or buy. Use at your own risk, it is amateur predictions, you have'
                    + ' been warned.<br />There will be also an application to track health metrics (weight,'
                    + ' % body fat, vascular fat ratio...). Or maybe not, this might be my next project.<br />'
                    + ' This project is open-sourced on <atarget=\'_blank\''
                    + ' href=\'https://github.com/jscoobyced/sawan.io\'>Github</a>. Enjoy and have fun.'
            },
            {
                article: 'Lorem Ipsum',
                articleTitle: this.data
            },
            {
                article: 'Some other news',
                articleTitle: this.data
            }
        ];
        return Promise.resolve({ articles });
    }
}
