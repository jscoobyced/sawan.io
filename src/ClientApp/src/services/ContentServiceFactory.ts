import { ContentService } from './ContentService';
import { IContentService } from './IContentService';
import { MockContentService } from './MockContentService';

export class ContentServiceFactory {

    public static GetContentService(mode: string): IContentService {
        if (mode === 'production') {
            return new ContentService();
        }

        return new MockContentService();
    }
}