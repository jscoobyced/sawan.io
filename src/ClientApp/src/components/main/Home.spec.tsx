import { shallow } from 'enzyme';
import React from 'react';
import { ContentService } from '../../services/ContentService';
import { Home } from './Home';

test('Home component is unchanged.', () => {
    const homeContent = new ContentService().defaultContent().homeContent;
    const home = shallow(<Home homeContent={homeContent} />);

    expect(home).toMatchSnapshot();
});

test('Home component with no article.', () => {
    const homeContent = new ContentService().defaultContent().homeContent;
    homeContent.article = [];
    const home = shallow(<Home homeContent={homeContent} />);

    expect(home).toMatchSnapshot();
});

test('Home component with no title.', () => {
    const homeContent = new ContentService().defaultContent().homeContent;
    homeContent.articleTitle = [];
    const home = shallow(<Home homeContent={homeContent} />);

    expect(home).toMatchSnapshot();
});

test('Home component with different lengths.', () => {
    const homeContent = new ContentService().defaultContent().homeContent;
    homeContent.articleTitle = [''];
    const home = shallow(<Home homeContent={homeContent} />);

    expect(home).toMatchSnapshot();
});

test('Home component with long article.', () => {
    const homeContent = new ContentService().defaultContent().homeContent;
    const longContent = 'x'.repeat(1024);
    homeContent.article = [longContent];
    const home = shallow(<Home homeContent={homeContent} />);

    expect(home).toMatchSnapshot();
});
