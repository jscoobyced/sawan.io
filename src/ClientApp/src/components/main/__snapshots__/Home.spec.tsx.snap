// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component is unchanged. 1`] = `
<article
  key="0"
>
  <h1 />
  <div>
    <span
      dangerouslySetInnerHTML={
        Object {
          "__html": "",
        }
      }
    />
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="link-article"
      to="/"
    >
      Read more...
    </NavLink>
  </div>
</article>
`;

exports[`Home component with different lengths. 1`] = `
<article
  key="0"
>
  <h1 />
  <div>
    <span
      dangerouslySetInnerHTML={
        Object {
          "__html": "",
        }
      }
    />
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="link-article"
      to="/"
    >
      Read more...
    </NavLink>
  </div>
</article>
`;

exports[`Home component with long article. 1`] = `
<article
  key="0"
>
  <h1 />
  <div>
    <span
      dangerouslySetInnerHTML={
        Object {
          "__html": "&#8230;",
        }
      }
    />
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="link-article"
      to="/"
    >
      Read more...
    </NavLink>
  </div>
</article>
`;

exports[`Home component with no article. 1`] = `<span />`;

exports[`Home component with no title. 1`] = `<span />`;
